modules/control/common/mrac_controller.cc

```diff
@@ -218,16 +219,38 @@ void MracController::BuildAdaptionModel() {
     matrix_b_adaption_(1, 0) = 1.0;
     matrix_p_adaption_(0, 0) = 1.0;
     matrix_p_adaption_(1, 1) = 1.0;
-    // TODO(Yu): check whether the high-order P matrix is positive definite
   } else {
     AWARN << "Adaption model order beyond the designed range, "
              "model_order: "
           << model_order_;
     adaption_model_enabled_ = false;
   }
+  if (matrix_a_reference_.rows() != matrix_p_adaption_.rows() ||
+      matrix_a_reference_.cols() != matrix_p_adaption_.cols()) {
+    AWARN << "Adaption matrix size does not match the reference matrix size";
+    adaption_model_enabled_ = false;
+  }
+  if (!CheckLyapunovPD(matrix_a_reference_, matrix_p_adaption_)) {
+    AWARN << "Solution of the algebraic Lyapunov equation is not symmetric "
+             "positive definite";
+    adaption_model_enabled_ = false;
+  }
   adaption_model_enabled_ = true;
 }
 
+bool MracController::CheckLyapunovPD(const Matrix matrix_a,
+                                     const Matrix matrix_p) const {
+  Matrix matrix_q = -matrix_p * matrix_a - matrix_a.transpose() * matrix_p;
+  Eigen::LLT<Matrix> llt_matrix_q(matrix_q);
+  // if matrix Q is not symmetric or the Cholkesky decomposition (LLT) failed
+  // due to the matrix Q are not positive definite
+  if (!matrix_q.isApprox(matrix_q.transpose()) ||
+      llt_matrix_q.info() == Eigen::NumericalIssue) {
+    return false;
+  }
```

Comment:
return (!matrix_q.isApprox(matrix_q.transpose()) ||
      llt_matrix_q.info() == Eigen::NumericalIssue)

Reply:
Refactor the original codes. Thanks.